### PR TITLE
Add physical context to examples, and use backend-agnostic functions

### PR DESCRIPTION
**Pull request by vishnu-2206**
As a physicist working with PINNs, I noticed that the 1D diffusion example lacked a physical description of the source term. 

I have : 
1. Added a detailed **Module Docstring** explaining the 1D Heat Equation context. 
2. Refactored the `pde` function to clearly identify the forcing term $f(x, t)$. 
3. Ensured that the source term is derived to satisfy the analytical solution $y = e^{-t} \sin(\pi x)$.

This makes the example more accessible for students and researchers using **DeepXDE** for scientific modeling.

### CODE REVIEW

File: examples/pinn_forward/diffusion_1d.py [hunk 1]
```diff
@@ -18,27 +28,34 @@ def pde(x, y):
...
+   
+   # Backend tensorflow.compat.v1 or tensorflow
     return (
         dy_t
         - dy_xx
-        + tf.exp(-x[:, 1:])
-        * (tf.sin(np.pi * x[:, 0:1]) - np.pi ** 2 * tf.sin(np.pi * x[:, 0:1]))
+        + source_term_val
     )
+
```
Comment (echen5503) [inline examples/pinn_forward/diffusion_1d.py]: Minimal changes, don't change any spaces not explicitly part of PR

File: examples/pinn_forward/diffusion_1d.py [hunk 2]
```diff
@@ -1,4 +1,14 @@
-"""Backend supported: tensorflow.compat.v1, tensorflow, pytorch, jax, paddle"""
+"""
+Backend supported: tensorflow.compat.v1, tensorflow, pytorch, jax, paddle
+1D Diffusion Equation with a Time-Dependent Source Term.
+
+This example solves the heat equation:
+∂y/∂t - ∂²y/∂x² = f(x, t)
+where the source term f(x, t) is chosen such that the analytical solution is y = e^(-t) * sin(πx).
+
+Physical context: This represents a 1D rod with a decaying heat source, where the ends are kept at zero temperature (Dirichlet Boundary Conditions).
```
Comment (echen5503) [inline examples/pinn_forward/diffusion_1d.py]: No need for this, this doesn't exist in other examples

File: examples/pinn_forward/diffusion_1d.py [hunk 3]
```diff
@@ -18,27 +28,34 @@ def pde(x, y):
     # Backend jax
     # dy_t, _ = dde.grad.jacobian(y, x, j=1)
     # dy_xx, _ = dde.grad.hessian(y, x, j=0)
-    # Backend tensorflow.compat.v1 or tensorflow
+    
+    # Physics Note: The following term is the forced heat source f(x, t)
+    # required to satisfy the analytical solution y = e^(-t)sin(πx).
+    source_term_val = tf.exp(-x[:, 1:]) * (tf.sin(np.pi * x[:, 0:1]) - np.pi ** 2 * tf.sin(np.pi * x[:, 0:1]))
```
Comment (echen5503) [inline examples/pinn_forward/diffusion_1d.py]: With this, note that you are making the tensorflow code different than the other backends. Also, note that this isn't the main point of the PR. If you really want to make this change, make it consistent for all backends, or use dde.backend.sin, dde.backend.exp, etc.

**Review by echen5503 — CHANGES_REQUESTED** [review]

**Comment by vishnu-2206** [comment]
Thank you for the guidance @echen5503. I have updated the PR to:
1. Revert the docstring and spacing to match the existing minimalist style
of the library.
2. Implement the source term using dde.backend to ensure consistency and
compatibility across all supported backends (TF, PyTorch, JAX, Paddle).
Let me know if any other adjustments are required.

File: examples/pinn_forward/diffusion_1d.py [hunk 4]
```diff
@@ -1,14 +1,4 @@
-"""
-Backend supported: tensorflow.compat.v1, tensorflow, pytorch, jax, paddle
-1D Diffusion Equation with a Time-Dependent Source Term.
-
-This example solves the heat equation:
-∂y/∂t - ∂²y/∂x² = f(x, t)
-where the source term f(x, t) is chosen such that the analytical solution is y = e^(-t) * sin(πx).
-
-Physical context: This represents a 1D rod with a decaying heat source, where the ends are kept at zero temperature (Dirichlet Boundary Conditions).
```
Comment (echen5503) [inline examples/pinn_forward/diffusion_1d.py]: I meant just remove L9, not remove the whole thing. The rest is useful.

**Review by echen5503 — CHANGES_REQUESTED** [review]

**Comment by echen5503** [comment]
@vishnu-2206 it should look like this
```
"""
Backend supported: tensorflow.compat.v1, tensorflow, pytorch, jax, paddle
1D Diffusion Equation with a Time-Dependent Source Term.

This example solves the heat equation:
∂y/∂t - ∂²y/∂x² = f(x, t)
where the source term f(x, t) is chosen such that the analytical solution is y = e^(-t) * sin(πx).
"""
import deepxde as dde
import numpy as np

def pde(x, y):
    # Most backends
    dy_t = dde.grad.jacobian(y, x, j=1)
    dy_xx = dde.grad.hessian(y, x, j=0)
    
    # Physics Note: The following term is the forced heat source f(x, t)
    # required to satisfy the analytical solution y = e^(-t)sin(πx).
    source_term_val = dde.backend.exp(-x[:, 1:]) * (dde.backend.sin(np.pi * x[:, 0:1]) - np.pi ** 2 * dde.backend.sin(np.pi * x[:, 0:1]))

    return ( 
        dy_t
        - dy_xx
        + source_term_val
    ) 

def func(x):
    return np.sin(np.pi * x[:, 0:1]) * np.exp(-x[:, 1:])


geom = dde.geometry.Interval(-1, 1)
timedomain = dde.geometry.TimeDomain(0, 1)
geomtime = dde.geometry.GeometryXTime(geom, timedomain)

bc = dde.icbc.DirichletBC(geomtime, func, lambda _, on_boundary: on_boundary)
ic = dde.icbc.IC(geomtime, func, lambda _, on_initial: on_initial)
data = dde.data.TimePDE(
    geomtime,
    pde,
    [bc, ic],
    num_domain=40,
    num_boundary=20,
    num_initial=10,
    solution=func,
    num_test=10000,
)

layer_size = [2] + [32] * 3 + [1]
activation = "tanh"
initializer = "Glorot uniform"
net = dde.nn.FNN(layer_size, activation, initializer)

model = dde.Model(data, net)

model.compile("adam", lr=0.001, metrics=["l2 relative error"])
losshistory, train_state = model.train(iterations=10000)

dde.saveplot(losshistory, train_state, issave=True, isplot=True)
```

**Comment by echen5503** [comment]
By only using deepXDE functions we can make examples backend-agnostic without casework.

File: examples/pinn_forward/diffusion_1d.py [hunk 5]
```diff
@@ -18,34 +18,16 @@ def pde(x, y):
     # Backend jax
     # dy_t, _ = dde.grad.jacobian(y, x, j=1)
     # dy_xx, _ = dde.grad.hessian(y, x, j=0)
-    # Backend tensorflow.compat.v1 or tensorflow
+    # Cross-backend source term
+    f = dde.backend.exp(-x[:, 1:]) * (
+        dde.backend.sin(np.pi * x[:, 0:1]) - np.pi**2 * dde.backend.sin(np.pi * x[:, 0:1])
+    )
+    # Backend tensorflow.compat.v1 or tensorflow, pytorch, jax, paddle
```
Comment (echen5503) [inline examples/pinn_forward/diffusion_1d.py]:
```suggestion
```

File: examples/pinn_forward/diffusion_1d_exactBC.py [hunk 6]
```diff
@@ -18,34 +18,16 @@ def pde(x, y):
     # Backend jax
     # dy_t, _ = dde.grad.jacobian(y, x, i=0, j=1)
     # dy_xx, _ = dde.grad.hessian(y, x, i=0, j=0)
-    # Backend tensorflow.compat.v1 or tensorflow
+    # Cross-backend source term
+    f = dde.backend.exp(-x[:, 1:]) * (
+        dde.backend.sin(np.pi * x[:, 0:1]) - np.pi**2 * dde.backend.sin(np.pi * x[:, 0:1])
+    )
+    # Backend tensorflow.compat.v1 or tensorflow, pytorch, jax, paddle
```
Comment (echen5503) [inline examples/pinn_forward/diffusion_1d_exactBC.py]:
```suggestion
```

File: examples/pinn_forward/diffusion_1d_resample.py [hunk 7]
```diff
@@ -18,36 +18,18 @@ def pde(x, y):
     # Backend jax
     # dy_t, _ = dde.grad.jacobian(y, x, i=0, j=1)
     # dy_xx, _ = dde.grad.hessian(y, x, i=0, j=0)
-    # Backend tensorflow.compat.v1 or tensorflow
+    # Cross-backend source term
+    f = dde.backend.exp(-x[:, 1:]) * (
+        dde.backend.sin(np.pi * x[:, 0:1]) - np.pi**2 * dde.backend.sin(np.pi * x[:, 0:1])
+    )
+    # Backend tensorflow.compat.v1 or tensorflow, pytorch, jax, paddle
```
Comment (echen5503) [inline examples/pinn_forward/diffusion_1d_resample.py]:
```suggestion
```

**Comment by echen5503** [comment]
It seems you missed some of my comments in the previous iterations. The point of the PR is to add documentation and keep the code concise with `dde.backend` functions. However, I requested you to add back the documentation and you did not do so.

Secondly, please focus on one file at a time. We can fix others in future PRs. Changing more and more files makes the PR hard to review.

Finally, please make sure to clean your AI output. AI will write comments to denote where things have changed, but it is not necessary here.

To reiterate, your code should look like this:
```
"""
Backend supported: tensorflow.compat.v1, tensorflow, pytorch, jax, paddle
1D Diffusion Equation with a Time-Dependent Source Term.

This example solves the heat equation:
∂y/∂t - ∂²y/∂x² = f(x, t)
where the source term f(x, t) is chosen such that the analytical solution is y = e^(-t) * sin(πx).
"""
import deepxde as dde
import numpy as np

def pde(x, y):
    # Most backends
    dy_t = dde.grad.jacobian(y, x, j=1)
    dy_xx = dde.grad.hessian(y, x, j=0)
    
    # Physics Note: The following term is the forced heat source f(x, t)
    # required to satisfy the analytical solution y = e^(-t)sin(πx).
    source_term_val = dde.backend.exp(-x[:, 1:]) * (dde.backend.sin(np.pi * x[:, 0:1]) - np.pi ** 2 * dde.backend.sin(np.pi * x[:, 0:1]))

    return ( 
        dy_t
        - dy_xx
        + source_term_val
    ) 

def func(x):
    return np.sin(np.pi * x[:, 0:1]) * np.exp(-x[:, 1:])


geom = dde.geometry.Interval(-1, 1)
timedomain = dde.geometry.TimeDomain(0, 1)
geomtime = dde.geometry.GeometryXTime(geom, timedomain)

bc = dde.icbc.DirichletBC(geomtime, func, lambda _, on_boundary: on_boundary)
ic = dde.icbc.IC(geomtime, func, lambda _, on_initial: on_initial)
data = dde.data.TimePDE(
    geomtime,
    pde,
    [bc, ic],
    num_domain=40,
    num_boundary=20,
    num_initial=10,
    solution=func,
    num_test=10000,
)

layer_size = [2] + [32] * 3 + [1]
activation = "tanh"
initializer = "Glorot uniform"
net = dde.nn.FNN(layer_size, activation, initializer)

model = dde.Model(data, net)

model.compile("adam", lr=0.001, metrics=["l2 relative error"])
losshistory, train_state = model.train(iterations=10000)

dde.saveplot(losshistory, train_state, issave=True, isplot=True)
```

In summary, please follow the format which I have described, and only change 1 file. Once this is merged, feel free to open a PR for the other 2.

**Comment by echen5503** [comment]
You can just copy-paste the code I provided verbatim. In future PRs, please keep my comments in mind.

**Comment by vishnu-2206** [comment]
Thank you for the guidance and the template. I have updated diffusion_1d.py
to use source_term_val and dde.backend to ensure it is fully
backend-agnostic as suggested.

Once you've had a chance to verify this structure, I will apply the exact
same logic to the other 1D diffusion examples (exactBC and resample) for
consistency. Ready for review!

**Comment by echen5503** [comment]
Thank you, this looks good. You can modify the other files to have the same structure.

**Comment by vishnu-2206** [comment]
Thank you for the feedback. Will modify the rest as well.

**Comment by echen5503** [comment]
LGTM. Care to add physics docs to 1d_resample as well?

**Comment by echen5503** [comment]
Once added, I think this will be ready to merge. Please update the title of the PR accordingly, to something like "Add physical context to examples, and use backend-agnostic functions".

**Comment by vishnu-2206** [comment]
I have updated the title and applied the same logic and structure to diffusion_1d_resample.py for consistency across the suite. These updates ensure physical accuracy and full compatibility with all DeepXDE backends by using dde.backend functions.

File: examples/pinn_forward/diffusion_1d_exactBC.py [hunk 8]
```diff
@@ -63,14 +50,8 @@ def func(x):
...
 net.apply_output_transform(
-    # Backend tensorflow.compat.v1 or tensorflow
-    lambda x, y: x[:, 1:2] * (1 - x[:, 0:1] ** 2) * y + tf.sin(np.pi * x[:, 0:1])
-    # Backend pytorch
-    # lambda x, y: x[:, 1:2] * (1 - x[:, 0:1] ** 2) * y + torch.sin(np.pi * x[:, 0:1])
-    # Backend jax
-    # lambda x, y: x[..., 1:2] * (1 - x[..., 0:1] ** 2) * y + jnp.sin(np.pi * x[..., 0:1])
-    # Backend paddle
-    # lambda x, y: x[:, 1:2] * (1 - x[:, 0:1] ** 2) * y + paddle.sin(np.pi * x[:, 0:1])
+    # This works for TensorFlow, PyTorch, JAX, and Paddle
```
Comment (echen5503) [inline examples/pinn_forward/diffusion_1d_exactBC.py]: No need to say this.

Comment (echen5503) [inline examples/pinn_forward/diffusion_1d_exactBC.py]: No need to say this.

File: examples/pinn_forward/diffusion_1d.py (hunk 5 above)
Comment (echen5503) [reply examples/pinn_forward/diffusion_1d.py]: No need to say this, because it is implied

**Review by echen5503 — CHANGES_REQUESTED** [review]
Just a small nitpick then we can merge this. Thanks for your help!

**Comment by vishnu-2206** [comment]
Thank you for the review! I've removed the redundant backend comments and
ensured a cleaner structure across diffusion_1d.py, exactBC.py, and
resample.py. Ready for the final merge!

**Comment by echen5503** [comment]
@lululxvi looks ready for merge. 

Please credit @vishnu2206 (first), and @echen5503 (second).

**Comment by vishnu-2206** [comment]
Thank you! Please credit @vishnu2206 as the primary author and @echen5503 for the guidance and review. I appreciate the help in getting these examples to the library standard!

**Comment by vishnu-2206** [comment]
I have amended the commit to include the co-author trailers as requested, crediting @vishnu-2206 (primary) and @echen5503 (secondary). The metadata should now be correctly reflected in the PR. Ready for final review.

**Comment by echen5503** [comment]
Hi, suggested code changes look good. But who is this Carrot-Bear account?

**Comment by vishnu-2206** [comment]
I have corrected my local Git configuration and amended the commit. The placeholder 'Carrot-Bear' has been replaced with my professional name, and the co-author metadata has been verified. The PR is now ready for your final review and merge.

**Comment by echen5503** [comment]
Looks good.
